core_lib/src/structure/layerbitmap.cpp

```diff
@@ -93,43 +92,56 @@ KeyFrame* LayerBitmap::createKeyFrame(int position, Object*)
     return b;
 }
 
-Status LayerBitmap::presave(const QString&)
+Status LayerBitmap::presave(const QString& sDataFolder)
 {
-    // handles those moved keys but note loaded yet
-    std::vector<BitmapImage*> bitmapArray;
-    foreachKeyFrame([&bitmapArray](KeyFrame* key)
+    QDir dataFolder(sDataFolder);
+    // Handles keys that have been moved but not modified
+    std::vector<BitmapImage*> movedOnlyBitmaps;
+    foreachKeyFrame([&movedOnlyBitmaps,&dataFolder,this](KeyFrame* key)
     {
         auto bitmap = static_cast<BitmapImage*>(key);
-        // null image + modified => the keyframe has been moved, but users didn't draw on it.
+        // (b->fileName() != fileName(b) && !modified => the keyframe has been moved, but users didn't draw on it.
         if (!bitmap->fileName().isEmpty()
-            && bitmap->image()->isNull() 
-            && bitmap->isModified())
+            && !bitmap->isModified()
+            && bitmap->fileName() != filePath(bitmap, dataFolder))
         {
-            bitmapArray.push_back(bitmap);
+            movedOnlyBitmaps.push_back(bitmap);
         }
     });
 
-    for (BitmapImage* b : bitmapArray) 
+    for (BitmapImage* b : movedOnlyBitmaps)
     {
-        if (b->fileName() != fileName(b))
-        {
-            QString tmpName = QString::asprintf("t_%03d.%03d.png", id(), b->pos());
+        // Move to temporary locations first to avoid overwritting anything we shouldn't be
+        // Ex: Frame A moves from 1 -> 2, Frame B moves from 2 -> 3. Make sure A does not overwrite B
+        QString tmpName = QString::asprintf("t_%03d.%03d.png", id(), b->pos());
+        QDir sA, sB;
+        if ((sA=QFileInfo(b->fileName()).dir()) != (sB=dataFolder)) {
+            // Copy instead of move if the data folder itself has changed
+            QFile::copy(b->fileName(), tmpName);
```

Comment:
Could you tell me in what scenario the data folder will change?

Reply:
When you load a pclx file, and save it as a pcl file, the data folder will change from the temporary directory to the .data directory. @Jose-Moreno first noted problems arising from this scenario which appear to be fixed now.